### PR TITLE
Don't initialize if using cpAsync

### PR DESCRIPTION
**Pull request by zasdfgbnm**
With the change in https://github.com/csarofeen/pytorch/pull/2484, `cp.async` will automatically fill zero for out of bound elements. So there is no need to zero-init any more.

This PR also contains some misc cleanup.

Example kernel:

```CUDA
__global__ void kernel1(Tensor<__half, 2> T0, Tensor<__half, 2> T1, Tensor<float, 2> T4) {
  alignas(16) extern __shared__ char array[];
  unsigned smem_offset = 0;
  int i594;
  i594 = (T0.size[1] * ((nvfuser_index_t)threadIdx.z)) * 16;
  int i597;
  i597 = (T0.size[1] * 8) * ((nvfuser_index_t)threadIdx.y);
  int i528;
  i528 = ((nvfuser_index_t)threadIdx.x) / 4;
  int i599;
  i599 = T0.size[1] * i528;
  int i548;
  i548 = ((nvfuser_index_t)threadIdx.x) % 4;
  int i601;
  i601 = i548 * 8;
  int i600;
  i600 = ((((T0.size[1] * ((nvfuser_index_t)blockIdx.y)) * 128) + i594) + i597) + i599;
  int i602;
  i602 = i600 + i601;
  int i605;
  i605 = T0.size[1] * 32;
  int i804;
  i804 = ((nvfuser_index_t)threadIdx.z) * 1024;
  int i806;
  i806 = 512 * ((nvfuser_index_t)threadIdx.y);
  int i808;
  i808 = 16 * ((nvfuser_index_t)threadIdx.x);
  int i1187;
  i1187 = ((((T0.size[1] * ((nvfuser_index_t)blockIdx.x)) * 128) + i594) + i597) + i599;
  int i1188;
  i1188 = i1187 + i601;
  int i1785;
  i1785 = (i600 + 64) + i601;
  int i2412;
  i2412 = (i1187 + 64) + i601;
  int i2819;
  i2819 = ((nvfuser_index_t)threadIdx.z) * 4096;
  int i2690;
  i2690 = ((nvfuser_index_t)threadIdx.x) / 8;
  int i2721;
  i2721 = i2690 % 2;
  int i2821;
  i2821 = i2721 * 512;
  int i2823;
  i2823 = 64 * (((nvfuser_index_t)threadIdx.x) % 8);
  int i2728;
  i2728 = i2690 / 2;
  int i2825;
  i2825 = i2728 * 16;
  int i3004;
  i3004 = ((nvfuser_index_t)threadIdx.y) * 4096;
  int i3017;
  i3017 = i2728 * 512;
  int i3020;
  i3020 = i2721 * 16;
  int i3988;
  i3988 = ((nvfuser_index_t)blockIdx.y) * 128;
  int i3990;
  i3990 = ((nvfuser_index_t)threadIdx.y) * 64;
  int i4009;
  i4009 = i548 * 2;
  int i4010;
  i4010 = ((((((T1.size[0] * ((nvfuser_index_t)blockIdx.x)) * 128) + ((T1.size[0] * ((nvfuser_index_t)threadIdx.z)) * 64)) + (T1.size[0] * i528)) + i3988) + i3990) + i4009;
  int i3994;
  i3994 = T1.size[0] * 16;
  int i4001;
  i4001 = T1.size[0] * 8;
  int i4656;
  i4656 = (8 * ((nvfuser_index_t)threadIdx.x)) + 7;
  int i4658;
  i4658 = (i4656 % 32) - T0.size[1];
  int i4679;
  i4679 = ((((nvfuser_index_t)threadIdx.z) * 16) + (8 * ((nvfuser_index_t)threadIdx.y))) + (i4656 / 32);
  int i4681;
  i4681 = (i4679 + i3988) - T1.size[0];
  int i5335;
  i5335 = ((nvfuser_index_t)blockIdx.x) * 128;
  int i5343;
  i5343 = (i4679 + i5335) - T0.size[0];
  int i6915;
  i6915 = (((((nvfuser_index_t)threadIdx.z) * 64) + i528) + i5335) - T0.size[0];
  int i6919;
  i6919 = ((i3990 + i4009) + i3988) - T1.size[0];
  smem_offset = alignBufferSize(smem_offset, 16);
  __half* T7 = reinterpret_cast<__half*>(array + smem_offset);
  smem_offset += (12288 * sizeof(__half));
  smem_offset = alignBufferSize(smem_offset, 16);
  __half* T6 = reinterpret_cast<__half*>(array + smem_offset);
  smem_offset += (12288 * sizeof(__half));
  unsigned i649;
  i649 = toSmem(T7);
  unsigned i809;
  i809 = ((i649 + i804) + i806) + i808;
  unsigned i1233;
  i1233 = toSmem(T6);
  unsigned i1400;
  i1400 = ((i1233 + i804) + i806) + i808;
  unsigned i2833;
  i2833 = ((i1233 + i2819) + i2821) + i2823;
  unsigned i2834;
  i2834 = i2833 + i2825;
  unsigned i3019;
  i3019 = ((i649 + i3004) + i3017) + i2823;
  unsigned i3021;
  i3021 = i3019 + i3020;
  unsigned i3230;
  i3230 = (i2833 + 32) + i2825;
  unsigned i3505;
  i3505 = (i3019 + 32) + i3020;
  float T5[128];
  #pragma unroll
  for(nvfuser_index_t i142 = 0; i142 < 4; ++i142) {
    int i234;
    i234 = 32 * i142;
    #pragma unroll
    for(nvfuser_index_t i143 = 0; i143 < 4; ++i143) {
      Ampere::initM16N16K16TN<16>(reinterpret_cast<Array<float,8,8>*>(&T5[(i234 + (4 * i143))]));
    }
  }
  #pragma unroll
  for(nvfuser_index_t i136 = 0; i136 < 2; ++i136) {
    int i603;
    i603 = 32 * i136;
    int i604;
    i604 = i602 + i603;
    int i810;
    i810 = 8192 * i136;
    unsigned i811;
    i811 = i809 + i810;
    int i1189;
    i1189 = i1188 + i603;
    unsigned i1401;
    i1401 = i1400 + i810;
    bool b4677;
    b4677 = i4658 < (-i603);
    #pragma unroll
    for(nvfuser_index_t i130 = 0; i130 < 4; ++i130) {
      Ampere::cpAsyncCg<__half, 8>((i811 + (2048 * i130)),reinterpret_cast<Array<__half,8,8>*>(&T1[(i604 + (i605 * i130))]),(b4677 && (i4681 < (-(32 * i130)))));
    }
    #pragma unroll
    for(nvfuser_index_t i119 = 0; i119 < 4; ++i119) {
      Ampere::cpAsyncCg<__half, 8>((i1401 + (2048 * i119)),reinterpret_cast<Array<__half,8,8>*>(&T0[(i1189 + (i605 * i119))]),(b4677 && (i5343 < (-(32 * i119)))));
    }
    Ampere::cpAsyncCommit();
  }
  Ampere::cpAsyncPartialBarrier<1>();
  __syncthreads();
  #pragma unroll 1
  for(nvfuser_index_t i137 = 0; i137 < (ceilDiv(T0.size[1], 32)); ++i137) {
    int i1779;
    i1779 = 32 * i137;
    int i1786;
    i1786 = i1785 + i1779;
    int i2024;
    i2024 = 8192 * ((2 + i137) % 3);
    unsigned i2028;
    i2028 = i809 + i2024;
    int i2413;
    i2413 = i2412 + i1779;
    unsigned i2655;
    i2655 = i1400 + i2024;
    int i2827;
    i2827 = 8192 * (i137 % 3);
    unsigned i2835;
    i2835 = i2834 + i2827;
    unsigned i3022;
    i3022 = i3021 + i2827;
    unsigned i3231;
    i3231 = i3230 + i2827;
    unsigned i3506;
    i3506 = i3505 + i2827;
    bool b6037;
    b6037 = (i4658 + i1779) < -64;
    #pragma unroll
    for(nvfuser_index_t i130 = 0; i130 < 4; ++i130) {
      Ampere::cpAsyncCg<__half, 8>((i2028 + (2048 * i130)),reinterpret_cast<Array<__half,8,8>*>(&T1[(i1786 + (i605 * i130))]),(b6037 && (i4681 < (-(32 * i130)))));
    }
    #pragma unroll
    for(nvfuser_index_t i119 = 0; i119 < 4; ++i119) {
      Ampere::cpAsyncCg<__half, 8>((i2655 + (2048 * i119)),reinterpret_cast<Array<__half,8,8>*>(&T0[(i2413 + (i605 * i119))]),(b6037 && (i5343 < (-(32 * i119)))));
    }
    Ampere::cpAsyncCommit();
    Array<__half, 64, 8> T8;
    Array<__half, 64, 8> T9;
    #pragma unroll
    for(nvfuser_index_t i121 = 0; i121 < 4; ++i121) {
      Turing::ldMatrix (*reinterpret_cast<Array<__half,8,8>*>(&T8[(8 * i121)]),(i2835 + (1024 * i121)));
    }
    #pragma unroll
    for(nvfuser_index_t i132 = 0; i132 < 4; ++i132) {
      Turing::ldMatrix (*reinterpret_cast<Array<__half,8,8>*>(&T9[(8 * i132)]),(i3022 + (1024 * i132)));
    }
    #pragma unroll
    for(nvfuser_index_t i140 = 0; i140 < 1; ++i140) {
      int i3225;
      i3225 = 32 * i140;
      unsigned i3232;
      i3232 = i3231 + i3225;
      int i3259;
      i3259 = 32 * ((1 + i140) % 2);
      int i3298;
      i3298 = 32 * (i140 % 2);
      unsigned i3507;
      i3507 = i3506 + i3225;
      #pragma unroll
      for(nvfuser_index_t i121 = 0; i121 < 4; ++i121) {
        Turing::ldMatrix (*reinterpret_cast<Array<__half,8,8>*>(&T8[(i3259 + (8 * i121))]),(i3232 + (1024 * i121)));
      }
      __half T2[32];
      #pragma unroll
      for(nvfuser_index_t i123 = 0; i123 < 4; ++i123) {
        int i3269;
        i3269 = 8 * i123;
        #pragma unroll
        for(nvfuser_index_t i125 = 0; i125 < 8; ++i125) {
          T2[(i3269 + i125)] = 0.00000000000000000e+00;
        }
      }
      #pragma unroll
      for(nvfuser_index_t i123 = 0; i123 < 4; ++i123) {
        int i3278;
        i3278 = 8 * i123;
        int i3302;
        i3302 = i3298 + i3278;
        #pragma unroll
        for(nvfuser_index_t i125 = 0; i125 < 8; ++i125) {
          T2[(i3278 + i125)]
             = T8[(i3302 + i125)];
        }
      }
      #pragma unroll
      for(nvfuser_index_t i132 = 0; i132 < 4; ++i132) {
        Turing::ldMatrix (*reinterpret_cast<Array<__half,8,8>*>(&T9[(i3259 + (8 * i132))]),(i3507 + (1024 * i132)));
      }
      __half T3[32];
      #pragma unroll
      for(nvfuser_index_t i134 = 0; i134 < 4; ++i134) {
        int i3545;
        i3545 = 8 * i134;
        #pragma unroll
        for(nvfuser_index_t i135 = 0; i135 < 8; ++i135) {
          T3[(i3545 + i135)] = 0.00000000000000000e+00;
        }
      }
      #pragma unroll
      for(nvfuser_index_t i134 = 0; i134 < 4; ++i134) {
        int i3554;
        i3554 = 8 * i134;
        int i3578;
        i3578 = i3298 + i3554;
        #pragma unroll
        for(nvfuser_index_t i135 = 0; i135 < 8; ++i135) {
          T3[(i3554 + i135)]
             = T9[(i3578 + i135)];
        }
      }
      #pragma unroll
      for(nvfuser_index_t i142 = 0; i142 < 4; ++i142) {
        int i3645;
        i3645 = 32 * i142;
        #pragma unroll
        for(nvfuser_index_t i143 = 0; i143 < 4; ++i143) {
          Ampere::M16N16K16TN<16>(
            reinterpret_cast<Array<float,8,8>*>(&T5[(i3645 + (4 * i143))]),
            &(reinterpret_cast<Array<__half,8,8>*>(&T2)[i142]),
            &(reinterpret_cast<Array<__half,8,8>*>(&T3)[i143]));
        }
      }
    }
    __half T2[32];
    #pragma unroll
    for(nvfuser_index_t i123 = 0; i123 < 4; ++i123) {
      int i3654;
      i3654 = 8 * i123;
      #pragma unroll
      for(nvfuser_index_t i125 = 0; i125 < 8; ++i125) {
        T2[(i3654 + i125)] = 0.00000000000000000e+00;
      }
    }
    #pragma unroll
    for(nvfuser_index_t i123 = 0; i123 < 4; ++i123) {
      int i3662;
      i3662 = 8 * i123;
      int i3681;
      i3681 = 32 + i3662;
      #pragma unroll
      for(nvfuser_index_t i125 = 0; i125 < 8; ++i125) {
        T2[(i3662 + i125)]
           = T8[(i3681 + i125)];
      }
    }
    __half T3[32];
    #pragma unroll
    for(nvfuser_index_t i134 = 0; i134 < 4; ++i134) {
      int i3689;
      i3689 = 8 * i134;
      #pragma unroll
      for(nvfuser_index_t i135 = 0; i135 < 8; ++i135) {
        T3[(i3689 + i135)] = 0.00000000000000000e+00;
      }
    }
    #pragma unroll
    for(nvfuser_index_t i134 = 0; i134 < 4; ++i134) {
      int i3697;
      i3697 = 8 * i134;
      int i3716;
      i3716 = 32 + i3697;
      #pragma unroll
      for(nvfuser_index_t i135 = 0; i135 < 8; ++i135) {
        T3[(i3697 + i135)]
           = T9[(i3716 + i135)];
      }
    }
    #pragma unroll
    for(nvfuser_index_t i142 = 0; i142 < 4; ++i142) {
      int i3780;
      i3780 = 32 * i142;
      #pragma unroll
      for(nvfuser_index_t i143 = 0; i143 < 4; ++i143) {
        Ampere::M16N16K16TN<16>(
          reinterpret_cast<Array<float,8,8>*>(&T5[(i3780 + (4 * i143))]),
          &(reinterpret_cast<Array<__half,8,8>*>(&T2)[i142]),
          &(reinterpret_cast<Array<__half,8,8>*>(&T3)[i143]));
      }
    }
    Ampere::cpAsyncPartialBarrier<1>();
    __syncthreads();
  }
  #pragma unroll
  for(nvfuser_index_t i148 = 0; i148 < 4; ++i148) {
    int i3846;
    i3846 = 32 * i148;
    int i4011;
    i4011 = i4010 + (i3994 * i148);
    int i6894;
    i6894 = -(16 * i148);
    #pragma unroll
    for(nvfuser_index_t i149 = 0; i149 < 4; ++i149) {
      int i3848;
      i3848 = i3846 + (4 * i149);
      int i3997;
      i3997 = 16 * i149;
      int i4012;
      i4012 = i4011 + i3997;
      int i6920;
      i6920 = -i3997;
      #pragma unroll
      for(nvfuser_index_t i150 = 0; i150 < 2; ++i150) {
        int i3850;
        i3850 = i3848 + (2 * i150);
        int i3999;
        i3999 = 8 * i150;
        int i4013;
        i4013 = i4012 + i3999;
        int i6921;
        i6921 = i6920 - i3999;
        #pragma unroll
        for(nvfuser_index_t i151 = 0; i151 < 2; ++i151) {
          int i3852;
          i3852 = i3850 + (16 * i151);
          int i4014;
          i4014 = i4013 + (i4001 * i151);
          bool b6916;
          b6916 = i6915 < (i6894 - (8 * i151));
          #pragma unroll
          for(nvfuser_index_t i152 = 0; i152 < 2; ++i152) {
            if ((b6916 && (i6919 < (i6921 - i152)))) {
              T4[(i4014 + i152)]
                 = T5[(i3852 + i152)];
            }
          }
        }
      }
    }
  }
}
```

### CODE REVIEW

File: third_party/nvfuser/runtime/memory.cu [hunk 1]
```diff
@@ -114,23 +114,7 @@ namespace Ampere {
 // Global to SMEM load that is asynchronous,
 // not guaranteed to be completed until cpAsyncBarrier() is called.
 template <typename dtype, int len>
-DEVICE_INLINE void cpAsync(unsigned smem_addr, void const* gmem_ptr) {
```
Comment (zasdfgbnm) [inline third_party/nvfuser/runtime/memory.cu:118]: This is not used, so I removed it.

File: third_party/nvfuser/runtime/memory.cu [hunk 2]
```diff
@@ -157,22 +141,6 @@ DEVICE_INLINE void cpAsync(
 // https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#cache-operators
 // not guaranteed to be completed until cpAsyncBarrier() is called.
 template <typename dtype, int len>
-DEVICE_INLINE void cpAsyncCg(unsigned smem_addr, void const* gmem_ptr) {
```
Comment (zasdfgbnm) [inline third_party/nvfuser/runtime/memory.cu:164]: This is not used, so I removed it.

File: third_party/nvfuser/csrc/kernel_ir.cpp [hunk 3]
```diff
@@ -157,6 +157,12 @@ Allocate::Allocate(
     TORCH_INTERNAL_ASSERT(alias->memoryType() == memory_type, "Invalid alias");
   }
 
+  // FIXME: there is a bug in lower_alias_memory.cpp that causes
+  // `NVFuserTest.FusionPredicateElimination6_CUDA` to fail if I simplify `5*2`
+  // into `10`
+
+  // size = simplifyExpr(size);
```
Comment (zasdfgbnm) [inline third_party/nvfuser/csrc/kernel_ir.cpp:164]: I am seeing this weird generated code:
```CUDA
    // Alias Allocation - register
    auto& T3 = T2;
    #pragma unroll
    for(nvfuser_index_t i38 = 0; i38 < 5; ++i38) {
      int i193;
      i193 = i192 + i38;
      #pragma unroll
      for(nvfuser_index_t i39 = 0; i39 < 2; ++i39) {
        T3[(i38 + (5 * i39))]
          = T2[(i193 + (3 * i39))]
          + (float) 1.00000000000000000e+00;
      }
    }
```

**Comment by zasdfgbnm** [comment]
cc @mmigdal-nv

File: third_party/nvfuser/csrc/kernel_ir.cpp (hunk 3 above)
Comment (naoyam) [reply third_party/nvfuser/csrc/kernel_ir.cpp:164]: What is this? What's the fusion math?

Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/kernel_ir.cpp:164]: It is a very simple pointwise fusion:
```C++
  Fusion fusion;
  FusionGuard fg(&fusion);

  auto tv0 = makeConcreteTensor({2, 3});
  fusion.addInput(tv0);

  auto tv1 = add(tv0, IrBuilder::create<Double>(1));
  auto tv2 = add(tv1, IrBuilder::create<Double>(1));
  auto tv3 = add(tv2, IrBuilder::create<Double>(1));
  auto tv4 = add(tv3, IrBuilder::create<Double>(1));
  fusion.addOutput(tv4);

  tv4->split(1, 5);
  TransformPropagatorWithCheck propagator(tv4);
  MaxRootDomainInfoSpanningTree(tv4).traverse(&propagator);

  tv4->reorder({{0, 1}, {1, 0}});
  tv3->computeAt(tv4, 1);
```

Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/kernel_ir.cpp:164]: CUDA kernel:
```CUDA
__global__ void kernel1(Tensor<float, 2> T0, Tensor<float, 2> T4) {
  NVFUSER_DEFINE_MAGIC_ZERO
  int i85;
  i85 = T0.stride[1] * 5;
  float T1[10];
  #pragma unroll
  for(nvfuser_index_t i32 = 0; i32 < 2; ++i32) {
    int i84;
    i84 = T0.stride[0] * i32;
    int i107;
    i107 = 3 * i32;
    #pragma unroll
    for(nvfuser_index_t i33 = 0; i33 < 1; ++i33) {
      int i87;
      i87 = i84 + (i85 * i33);
      int i108;
      i108 = 5 * i33;
      int i109;
      i109 = i107 + i108;
      int i270;
      i270 = 3 - i108;
      #pragma unroll
      for(nvfuser_index_t i34 = 0; i34 < 5; ++i34) {
        int i56;
        i56 = i34 + nvfuser_zero;
        if ((i56 < i270)) {
          T1[(i109 + i34)]
            = T0[(i87 + (T0.stride[1] * i56))]
            + (float) 1.00000000000000000e+00;
        }
      }
    }
  }
  NVFUSER_UPDATE_MAGIC_ZERO
  float T2[10];
  #pragma unroll
  for(nvfuser_index_t i35 = 0; i35 < 2; ++i35) {
    int i128;
    i128 = 3 * i35;
    #pragma unroll
    for(nvfuser_index_t i36 = 0; i36 < 1; ++i36) {
      int i130;
      i130 = i128 + (5 * i36);
      #pragma unroll
      for(nvfuser_index_t i37 = 0; i37 < 5; ++i37) {
        T2[(i130 + i37)] = 0;
      }
    }
  }
  NVFUSER_UPDATE_MAGIC_ZERO
  #pragma unroll
  for(nvfuser_index_t i35 = 0; i35 < 2; ++i35) {
    int i151;
    i151 = 3 * i35;
    #pragma unroll
    for(nvfuser_index_t i36 = 0; i36 < 1; ++i36) {
      int i152;
      i152 = 5 * i36;
      int i153;
      i153 = i151 + i152;
      int i304;
      i304 = 3 - i152;
      #pragma unroll
      for(nvfuser_index_t i37 = 0; i37 < 5; ++i37) {
        int i154;
        i154 = i153 + i37;
        if (((i37 + nvfuser_zero) < i304)) {
          T2[i154]
            = T1[i154]
            + (float) 1.00000000000000000e+00;
        }
      }
    }
  }
  NVFUSER_UPDATE_MAGIC_ZERO
  #pragma unroll
  for(nvfuser_index_t i40 = 0; i40 < 1; ++i40) {
    int i192;
    i192 = 5 * i40;
    int i338;
    i338 = 3 - i192;
    // Alias Allocation - register
    auto& T3 = T2;
    #pragma unroll
    for(nvfuser_index_t i38 = 0; i38 < 5; ++i38) {
      int i193;
      i193 = i192 + i38;
      #pragma unroll
      for(nvfuser_index_t i39 = 0; i39 < 2; ++i39) {
        T3[(i38 + (5 * i39))]
          = T2[(i193 + (3 * i39))]
          + (float) 1.00000000000000000e+00;
      }
    }
    #pragma unroll
    for(nvfuser_index_t i41 = 0; i41 < 2; ++i41) {
      int i212;
      i212 = 5 * i41;
      int i237;
      i237 = i192 + (3 * i41);
      #pragma unroll
      for(nvfuser_index_t i42 = 0; i42 < 5; ++i42) {
        int i214;
        i214 = i42 + nvfuser_zero;
        if ((i214 < i338)) {
          T4[(i237 + i214)]
            = T3[(i212 + i42)]
            + (float) 1.00000000000000000e+00;
        }
      }
    }
  }
  NVFUSER_UPDATE_MAGIC_ZERO
}
```

Comment (naoyam) [reply third_party/nvfuser/csrc/kernel_ir.cpp:164]: Does it produce wrong results?

Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/kernel_ir.cpp:164]: Yes, it produces wrong result. Let's move the discussion of this issue here: https://github.com/csarofeen/pytorch/pull/2490

Comment (naoyam) [reply third_party/nvfuser/csrc/kernel_ir.cpp:164]: Ah, the traversal pattern is different between `T3` and `T2`, so they can't be reused.

Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/kernel_ir.cpp:164]: Yes, correct.

File: third_party/nvfuser/csrc/kernel_ir.cpp [hunk 4]
```diff
@@ -182,11 +188,7 @@ Allocate::Allocate(
           buffer,
           memory_type,
           size == nullptr ? std::vector<Val*>{} : std::vector<Val*>{size},
-          zero_init) {
-  TORCH_INTERNAL_ASSERT(
```
Comment (naoyam) [inline third_party/nvfuser/csrc/kernel_ir.cpp:186]: Why is this removed?

File: third_party/nvfuser/csrc/lower_allocation.cpp [hunk 5]
```diff
@@ -480,10 +481,18 @@ class AllocationInserter : public kir::ExprMutator {
...
-      } else if (default_val != nullptr) {
+      } else {
         init = default_val;
       }
 
+      if (ir_utils::isCpAsyncOp(expr)) {
+        TORCH_CHECK(
+            init == nullptr || init->isZero(),
+            "cp.async initialized with non-zero is not supported");
+        // cp.async will automatically fill zero when out of bound
```
Comment (naoyam) [inline third_party/nvfuser/csrc/lower_allocation.cpp:492]: Interesting. So, does that mean if the source buffer on global memory is accessed beyond its buffer limit, `cp.async` automatically returns zero?

Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/lower_allocation.cpp:492]: Yes. (assuming that your predicate is correct)

File: third_party/nvfuser/csrc/kernel_ir.cpp (hunk 4 above)
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/kernel_ir.cpp:186]: Because this ctor calls another ctor, and that ctor will do this check.

**Review by naoyam — APPROVED** [review]
LGTM
